linode/networkingip/framework_resource.go

```diff
@@ -0,0 +1,178 @@
+	var state NetworkingIPModel
+
+	resp.Diagnostics.Append(req.State.Get(ctx, &state)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	client := r.Meta.Client
+
+	ip, err := client.GetIPAddress(ctx, state.ID.ValueString())
+	if err != nil {
+		if lerr, ok := err.(*linodego.Error); ok && lerr.Code == 404 {
+			resp.State.RemoveResource(ctx)
+			return
+		}
+		resp.Diagnostics.AddError(
+			"Error reading IP Address",
+			fmt.Sprintf("Could not read IP address %s: %s", state.ID.ValueString(), err),
+		)
+		return
+	}
+
+	state.FlattenIPAddress(ctx, ip, true)
+	resp.Diagnostics.Append(resp.State.Set(ctx, state)...)
+}
+
+func (r *Resource) Update(ctx context.Context, req resource.UpdateRequest, resp *resource.UpdateResponse) {
+	tflog.Debug(ctx, "Update linode_networking_ip")
+	var plan, state NetworkingIPModel
+
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &plan)...)
+	resp.Diagnostics.Append(req.State.Get(ctx, &state)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	client := r.Meta.Client
+
+	var reservedValue *bool
+	if !plan.Reserved.IsNull() {
+		value := plan.Reserved.ValueBool()
+		reservedValue = &value
+	}
+
+	updateOpts := linodego.IPAddressUpdateOptions{
+		Reserved: reservedValue,
+	}
+
+	ip, err := client.UpdateIPAddress(ctx, state.Address.ValueString(), updateOpts)
+	if err != nil {
+		resp.Diagnostics.AddError(
+			"Failed to Update IP Address",
+			fmt.Sprintf("Could not update reserved status of IP address: %s", err),
+		)
+		return
+	}
+
+	plan.FlattenIPAddress(ctx, ip, false)
+	resp.Diagnostics.Append(resp.State.Set(ctx, &plan)...)
+}
+
+func (r *Resource) Delete(ctx context.Context, req resource.DeleteRequest, resp *resource.DeleteResponse) {
+	tflog.Debug(ctx, "Delete linode_networking_ip")
+	var state NetworkingIPModel
+
+	resp.Diagnostics.Append(req.State.Get(ctx, &state)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	client := r.Meta.Client
+
+	// For created/reserved IPs
+	if !state.Reserved.ValueBool() {
+		// This is a regular IP address
+		linodeID := helper.FrameworkSafeInt64ToInt(state.LinodeID.ValueInt64(), &resp.Diagnostics)
+		if resp.Diagnostics.HasError() {
+			return
+		}
+
+		err := client.DeleteInstanceIPAddress(ctx, linodeID, state.Address.ValueString())
+		if err != nil {
+			if lErr, ok := err.(*linodego.Error); (ok && lErr.Code != 404) || !ok {
+				resp.Diagnostics.AddError(
+					"Failed to Delete IP",
+					fmt.Sprintf(
+						"failed to delete instance (%d) ip (%s): %s",
+						linodeID, state.Address.ValueString(), err.Error(),
+					),
+				)
+			}
+		}
+	} else {
+		// This is a reserved IP address
+		err := client.DeleteReservedIPAddress(ctx, state.Address.ValueString())
+		if err != nil {
+			if lErr, ok := err.(*linodego.Error); (ok && lErr.Code != 404) || !ok {
+				resp.Diagnostics.AddError(
+					"Failed to Delete Reserved IP",
+					fmt.Sprintf(
+						"failed to delete reserved ip (%s): %s",
+						state.Address.ValueString(), err.Error(),
+					),
+				)
+			}
```

Comment:
Just a heads up, I noticed a 404 is suppressed here when using `v4` instead of `v4beta`. The 404 checking logic is necessary so there's not much we can do to remedy this without manually checking the API version on the client.

Reply:
Weirdly enough tainting a reserved IP seems to not properly destroy the IP even when the DELETE endpoint returns a 200 🤔

Reply:
Ohh I see, deleting a reserved IP seems to just convert it to ephemeral. In that case I think this logic should:

1. Always call `client.DeleteInstanceIPAddress(...)`
2. Only call `client.DeleteReservedIPAddress(...)` if the IP is a reserved IP

Reply:
I agree. This was the conditional delete that we discussed earlier in the huddle. I will make the change.